src/cmd/api/controllers/video_upload_handler.go

```diff
@@ -160,125 +155,146 @@ func (v VideoUploadHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		log.Error("Not supported file type : " + mime.String() + " (" + mime.Extension() + ")")
 		return false
 	}
 
-	for _, ext := range supportedExtention {
+	for _, ext := range []string{".jpeg", ".jpg", ".png"} {
 		if strings.EqualFold(mime.Extension(), ext) {
 			return true
 		}
 	}
+
 	log.Error("Not supported file type : " + mime.String() + " (" + mime.Extension() + ")")
 	return false
 }
 
-func (v VideoUploadHandler) checkCreateVideo(ctx context.Context, w http.ResponseWriter, title string) (*models.Video, int) {
-	// Check if the returned error comes from duplicate title
-	videoCreated, err := v.VideosDAO.GetVideoFromTitle(ctx, title)
-	if err != nil {
-		log.Error("Cannot find video ", title, "  : ", err)
-		log.Error("Cannot insert new video into database: ", err)
-		w.WriteHeader(http.StatusInternalServerError)
-		return videoCreated, http.StatusInternalServerError
-	}
+func (v VideoUploadHandler) recoverAfterError(ctx context.Context, video *models.Video, fileCover, fileVideo multipart.File, fileHandler *multipart.FileHeader, w http.ResponseWriter) {
+	// Here, it means that the video already exists, check for its status
+	if video.Status == models.FAIL_UPLOAD {
+		log.Debug("Try to re-upload and re-encode failed video")
+		coverPath, err := v.uploadCover(ctx, fileCover, video.ID, fileHandler)
+		if err != nil {
+			log.Error("Cannot upload cover image : ", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
 
-	log.Debug("This title already exist, check video status")
-	if videoCreated.Status == models.FAIL_UPLOAD {
-		// Retry to upload+encode
-		log.Debugf("Last upload of video %v failed, simply retry", videoCreated.Title)
+		videoCreated, err := v.uploadVideo(ctx, video.ID, video.Title, video.SourcePath, coverPath, fileVideo, video)
+		if err != nil {
+			log.Error("Cannot upload video : ", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
 
-	} else if videoCreated.Status == models.FAIL_ENCODE {
-		// Retry to encode
-		log.Debug("Ask for video encoding")
-		metrics.CounterVideoEncodeRequest.Inc()
 		if err = v.sendVideoForEncoding(ctx, videoCreated); err != nil {
-			metrics.CounterVideoEncodeFail.Inc()
 			log.Error("Cannot send video for encoding : ", err)
 			w.WriteHeader(http.StatusInternalServerError)
-			return videoCreated, http.StatusInternalServerError
+			return
 		}
 
-		// Here, everything went well: video encoding asked and video status updated
-		return videoCreated, http.StatusOK
+		// Include videoCreated and HATEOAS upload link into response
+		writeHTTPResponse(videoCreated, w)
+		log.Infof("Video '%v' successfully uploaded", video.Title)
+		return
 
-	} else {
-		// Title already exist, video already upload and encode, return error
-		log.Error("A video with this title already uploaded and encoded")
-		http.Error(w, "This title already exists", http.StatusConflict)
-		return videoCreated, http.StatusConflict
+	} else if video.Status == models.FAIL_ENCODE {
```

Comment:
Since we only use this function if models.status is FAIL_ENCODE or FAIL_UPLOAD, is it necessary to duplicate this code ? We could get this part out of the "if" if we replace 'videoCreated' with 'video'.